graph/api/src/test/scala/org/apache/spark/graph/api/PropertyGraphSuite.scala

```diff
@@ -0,0 +1,309 @@
+      .toDF(
+        ID_COLUMN,
+        label("Person"),
+        label("Student"),
+        label("Teacher"),
+        "age",
+        "name",
+        "subject")
+
+    val expectedRelDf = spark
+      .createDataFrame(
+        Seq(
+          (convertId(0L), convertId(0L), convertId(1L), true, false, Some(1984)),
+          (convertId(1L), convertId(2L), convertId(1L), false, true, None)))
+      .toDF(
+        ID_COLUMN,
+        SOURCE_ID_COLUMN,
+        TARGET_ID_COLUMN,
+        label("KNOWS"),
+        label("TEACHES"),
+        "since")
+
+    checkAnswer(graph.nodes, expectedNodeDf)
+    checkAnswer(graph.relationships, expectedRelDf)
+  }
+
+  test("create graph with explicit property-to-column mappings") {
+    val studentDF = spark
+      .createDataFrame(Seq((0L, "Alice", 42), (1L, "Bob", 23)))
+      .toDF("id", "col_name", "col_age")
+    val teacherDF = spark
+      .createDataFrame(Seq((2L, "Eve", "CS")))
+      .toDF("id", "col_name", "col_subject")
+
+    val studentNF = NodeFrame(
+      studentDF,
+      "id",
+      Set("Person", "Student"),
+      properties = Map("name" -> "col_name", "age" -> "col_age"))
+    val teacherNF = NodeFrame(
+      teacherDF,
+      "id",
+      Set("Person", "Teacher"),
+      properties = Map("name" -> "col_name", "subject" -> "col_subject"))
+
+    val knowsDF = spark
+      .createDataFrame(Seq((0L, 0L, 1L, 1984)))
+      .toDF("id", "source", "target", "col_since")
+    val teachesDF = spark.createDataFrame(Seq((1L, 2L, 1L))).toDF("id", "source", "target")
+
+    val knowsRF = RelationshipFrame(
+      knowsDF,
+      "id",
+      "source",
+      "target",
+      relationshipType = "KNOWS",
+      properties = Map("since" -> "col_since"))
+    val teachesRF = RelationshipFrame(
+      teachesDF,
+      "id",
+      "source",
+      "target",
+      "TEACHES",
+      Map.empty)
+
+    val graph = cypherSession.createGraph(Array(studentNF, teacherNF), Array(knowsRF, teachesRF))
+
+    val expectedNodeDf = spark
+      .createDataFrame(
+        Seq(
+          (convertId(0L), true, true, false, Some(42), Some("Alice"), None),
+          (convertId(1L), true, true, false, Some(23), Some("Bob"), None),
+          (convertId(2L), true, false, true, None, Some("Eve"), Some("CS"))))
+      .toDF(
+        ID_COLUMN,
+        label("Person"),
+        label("Student"),
+        label("Teacher"),
+        "age",
+        "name",
+        "subject")
+
+    val expectedRelDf = spark
+      .createDataFrame(
+        Seq(
+          (convertId(0L), convertId(0L), convertId(1L), true, false, Some(1984)),
+          (convertId(1L), convertId(2L), convertId(1L), false, true, None)))
+      .toDF(
+        ID_COLUMN,
+        SOURCE_ID_COLUMN,
+        TARGET_ID_COLUMN,
+        label("KNOWS"),
+        label("TEACHES"),
+        "since")
+
+    checkAnswer(graph.nodes, expectedNodeDf)
+    checkAnswer(graph.relationships, expectedRelDf)
+  }
+
+  lazy val nodes: DataFrame = spark
```

Comment:
Let's move these `val nodes` and `val relationships` to the top of this test suite.